projects/wallet-lib-client-for-event-queue/0001-event-queue-connection-design.md

```diff
@@ -0,0 +1,199 @@
+
+## Events
+
+The events from the event queue are not filtered in any way, meaning we have to implement our own filter.
+The events that represent new transactions or blocks is called `NEW_VERTEX_ACCEPTED` and any updates on the transaction data is called `VERTEX_METADATA_CHANGED` these 2 events come with the current transaction data, we can use this data to filter for transaction of our wallet.
+
+## Address subscription
+
+The current implementation uses the fullnode pubsub to listen only for addresses of our wallet, meaning that during startup we send a subscribe command for each address of our wallet.
+Since the events will be filtered locally we can instead use the `subscribeAddresses` method to create a local list of addresses being listened to and use this list to filter the events.
+
+The "list of addresses" will be an object where the address is the key, since determining if an object has a key is O(1) we can ensure that this does not become a bottleneck for wallets with many addresses.
+Alternatively, we could use the storage `isAddressMine` method which is already O(1).
+
+## Event streaming
+
+To get the full state of the wallet we would need to stream all events of the fullnode, but we can still use the address history api to fetch the balance and transactions of our addresses and start listening the newer events.
+
+The best way to achieve this is to use the event api to fetch a single event, this will come with the latest event id.
+Example response of `GET ${FULLNODE_URL}/v1a/event?size=1`
+
+```json
+{
+    "events": [
+        {
+            "peer_id": "ca084565aa4ac6f84c452cb0085c1bc03e64317b64e0761f119b389c34fcfede",
+            "id": 0,
+            "timestamp": 1686186579.306944,
+            "type": "LOAD_STARTED",
+            "data": {},
+            "group_id": null
+        }
+    ],
+    "latest_event_id": 9038
+}
+```
+
+Then we save `latest_event_id` and sync the history with the address history api.
+Once we have the wallet on the current state we can start streaming from the `latest_event_id`.
+There can be transactions arriving during this process which would mean we add them during the history sync and during the event streaming, but this issue does not affect the balance or how we process the history.
+
+## Best block update
+
+The fullnode pubsub sends updates whenever the best chain height changes, this is so our wallets can unlock our block rewards, the event queue does not send an update like this but we receive all block transactions as events, meaning we can listen for any transaction with `version` 0 or 3 (block or merged mining block) and with the metadata `voided_by` as `null` (this is because if a block is not voided, it is on the main chain) and derive the best chain height.
+
+We will always expect the latest unvoided block to be the best chain newest block since during re-orgs (where the best chain changes) we will receive updates and the new best chain will be updated with `voided_by` as `null`.
```

Comment:
We have the `current heigth` saved. If a valid and unvoided block arrives, we update the `current height`.
This works with `NEW_VERTEX_ACCEPTED` and `VERTEX_METADATA_CHANGED`.

The `voided_by` filter is just a precaution against branches on the blockchain, since the only unvoided blocks are the best chain blocks.
This also works well with re-orgs since we receive a `VERTEX_METADATA_CHANGED` to void a block but also receive one to unvoid the new best chain block, which would update the height accordingly.